data/tutorial/architecture.md

```diff
@@ -18,6 +18,14 @@ It is also a **content addressable** store as _values_ (be it contents, commits
 or nodes) are stored as a pair `(hash(value), value)` and therefore a value can
 be accessed using its hash.
 
+A collection of block stores is available in Irmin, but you can create your own
+using the interface [Irmin.CONTENT_ADDRESSABLE_STORE] as long as the following
+are guaranteed:
```

Comment:
These seem like odd constraints to specify. Isn't it inherently true that any reads of a single key will return the same value? (since the value is content-addressed). Similarly, aren't all reads of a single key going to return the unique value that is addressed by that key?

Is this about hash collisions? If so, why specify _concurrent_ reads / writes?

Reply:
It's not about hash collisions. I was trying to preserve the structure of https://github.com/mirage/irmin/wiki/Concurrency, which maybe has a better phrasing. 

The point is to specify the behaviour in case of concurrent reads/writes, which agree, does not make a lot of sense for the content store.  But you want to prevent the content store to be a random number generator let's say, and this is up to the user to enforce it.

Reply:
I agree that one needs to enforce such constraints, but think they should be tightened to require _all_ writes of the same value to return the same key, not only those that are concurrent (and similarly for reads).

In the case of the block store, I feel that this is not really a concurrency assumption so much as a 'satisfies content-addressed semantics' assumption.